### PR TITLE
Add recipe for ckzg

### PR DESCRIPTION
**Pull request by step21**
Checklist
- [x] Title of this PR is meaningful: e.g. "Adding my_nifty_package", not "updated meta.yaml".
- [x] License file is packaged (see [here](https://github.com/conda-forge/staged-recipes/blob/5eddbd7fc9d1502169089da06c3688d9759be978/recipes/example/meta.yaml#L64-L73) for an example).
- [x] Source is from official source.
- [x] Package does not vendor other packages. (If a package uses the source of another package, they should be separate packages or the licenses of all packages need to be packaged).
- [x] If static libraries are linked in, the license of the static library is packaged.
- [x] Package does not ship static libraries. If static libraries are needed, [follow CFEP-18](https://github.com/conda-forge/cfep/blob/main/cfep-18.md).
- [x] Build number is 0.
- [x] A tarball (`url`) rather than a repo (e.g. `git_url`) is used in your recipe (see [here](https://conda-forge.org/docs/maintainer/adding_pkgs.html#build-from-tarballs-not-repos) for more details).
- [x] GitHub users listed in the maintainer section have posted a comment confirming they are willing to be listed there.
- [x] When in trouble, please check our [knowledge base documentation](https://conda-forge.org/docs/maintainer/knowledge_base.html) before pinging a team.

### CODE REVIEW

**Comment by zklaus** [comment]
Note that upstream has in the meantime added a patch for the Windows issue at https://github.com/ethereum/c-kzg-4844/pull/419. You can turn that into a patch here, or skip Windows until a new release with the patch comes out.

**Comment by step21** [comment]
> Note that upstream has in the meantime added a patch for the Windows issue at [ethereum/c-kzg-4844#419](https://github.com/ethereum/c-kzg-4844/pull/419). You can turn that into a patch here, or skip Windows until a new release with the patch comes out.

I saw this, because as you can see I was in that discussion with upstream, but there is no indication that that patch solves this particular windows issue. I had wanted to test it, but as I wrote had/have trouble getting a system able to test it.

File: recipes/ckzg/meta.yaml
Comment (xhochy) [inline recipes/ckzg/meta.yaml:21]: Why is `clang` listed here? You probably would either list it as a compiler or use `clangdev`.

Comment (step21) [reply recipes/ckzg/meta.yaml]: thanks. Most likely because it worked, and I didn't know that I could add it with compiler(). Interestingly, doing that meant it failed on Windows. So now I changed it to that but only use that line for `[lin]`.

File: recipes/ckzg/meta.yaml
Comment (xhochy) [inline recipes/ckzg/meta.yaml:20]: A cleaner approach for this is described in https://conda-forge.org/news/2024/04/30/clang-everywhere/ Havig both `compiler('c')` and `compiler('clang')` can lead to issues. Putting clang into cbc.yaml should prevent this.

Comment (step21) [reply recipes/ckzg/meta.yaml]: I tried that but it fails unfortunately. This might be because the build script explicitly uses gcc for windows. https://github.com/step21/c-kzg-4844/blob/main/src/Makefile

Comment (xhochy) [reply recipes/ckzg/meta.yaml]: Best is to limit it with `  # [linux]` on all lines of the cbc.yaml

Comment (step21) [reply recipes/ckzg/meta.yaml]: Somehow this is not working. Error from OSX `  /var/folders/n8/w901w1rn4ldg6yskstndsj200000gn/T/assembly-f95753.s:6555:1: error: invalid CFI advance_loc expression
  .cfi_adjust_cfa_offset 8`
Windows build just tries to apply clang options to gcc, which fails ofc.

Comment (step21) [reply recipes/ckzg/meta.yaml]: @xhochy  Please advise if I should fix/remove the build script from upstream to enable clang for all (assuming it works) or revert to the version that worked using the built-in build script, but uses gcc for windows.

**Comment by step21** [comment]
@conda-forge/core please review/approve with the original build script or tell me how I can force/replace the build script to use only clang. The way @xhochy suggested did not work, thus I reverted the commits to a working build.

**Comment by hmaarrfk** [comment]
what if you removed clang alltogether?

**Comment by step21** [comment]
> what if you removed clang alltogether?

I assume it would still clash with this Makefile https://github.com/ethereum/c-kzg-4844/blob/main/src/Makefile which has 
```
# Cross-platform compilation settings.
ifeq ($(PLATFORM),Windows)
	CC = gcc
	CFLAGS += -D_CRT_SECURE_NO_WARNINGS
else
	CC = clang
	CFLAGS += -fPIC -Werror
endif
```
Windows also probably does not really use this file, as in `setup.py` there is

```
class CustomBuild(build_ext):
    def run(self):
        if system() == "Windows":
            try:
                check_call(["blst\\build.bat"])
            except Exception:
                pass
        check_call(["make", "-C", "src", "blst"])
        super().run()
```

I'm not saying these are great, but I assume I would have to patch them only use either clang or gcc. (and I do not know if there was a reason to only use either)

**Comment by isuruf** [comment]
you'll have to patch `setup.py` to use something like
```
check_call(["make", "-C", "src", "blst", f"CC={os.environ['CC']}"])
```

**Comment by step21** [comment]
This unfortunately fails on Win, which still uses cl.exe. Any ideas? (I also patched out the use of the windows bat file, as it didn't use either gcc or clang and I'm not sure what kind of magic it is doing.
https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=961378&view=logs&jobId=240f1fee-52bc-5498-a14a-8361bde76ba0&j=240f1fee-52bc-5498-a14a-8361bde76ba0&t=7c0f8eae-6d6f-51bf-636f-73a1a7fb1bc4

**Comment by jakirkham** [comment]
Maybe disable Windows and then we can review/merge?

The Windows portion could be revisited in the feedstock

**Comment by step21** [comment]
> Maybe disable Windows and then we can review/merge?
> 
> The Windows portion could be revisited in the feedstock

Ok good idea. Windows build is skipped now.

**Comment by step21** [comment]
This is ready for review, @conda-forge/help-python

File: recipes/ckzg/meta.yaml
Comment (jakirkham) [inline recipes/ckzg/meta.yaml:12]: Let's drop this patch if unused. It can be attempted again in the feedstock

```suggestion
```

File: recipes/ckzg/build.patch
Comment (jakirkham) [inline recipes/ckzg/build.patch:18]:
```suggestion
```

File: recipes/ckzg/meta.yaml
Comment (jakirkham) [reply recipes/ckzg/meta.yaml:24]: This should be unneeded as we have `{{ compiler('c') }}` already. Let's drop it

```suggestion
```

**Review by jakirkham — COMMENTED** [review]
Thanks Florian! 🙏

Added a couple minor suggestions below

File: recipes/ckzg/meta.yaml
Comment (jakirkham) [inline recipes/ckzg/meta.yaml:24]: This should be unneeded as we have `{{ compiler('c') }}` already. Let's drop it

```suggestion
```

File: recipes/ckzg/meta.yaml
Comment (jakirkham) [inline recipes/ckzg/meta.yaml:9]:
```suggestion
  url: https://pypi.io/packages/source/{{ name[0] }}/{{ name }}/{{ name }}-{{ version }}.tar.gz
```

**Comment by step21** [comment]
Hey, @jakirkham without any patch, we are now again at the juncture of this Makefile I referred to here: https://github.com/conda-forge/staged-recipes/pull/26068#issuecomment-2181023138
This forces clang for linux and osx. On OSX that works, because it is installed by default. The main question is - how to force it to use only one compiler? In the pursuit of first merging, then fixing the feedstock, should I add clang back, so that only clang is used?

**Comment by MementoRC** [comment]
@step21 After a few days looking into it, I proposed this: https://github.com/conda-forge/staged-recipes/pull/27412, which should build Windows using gcc 5, separate `blst` as its own package (with the py-binding being worked in a separate recipe that still has a few snags on linux)
Note that it is still work in progress as I need to fine-tune windows/osx by going through CI.

**Comment by MementoRC** [comment]
Cool. I am still working on it. I want to enforce linking to the versioned dynamic library and it's a bit trickier. It feels more reliable, but maybe overkill. As for windows, it's my Achilles heel, I had it working but must have changed something during the many refactoring. I'll ping you when I get it to pass

**Comment by MementoRC** [comment]
Similar functionality with separated `blst`/`ckzg` build to address the upstream package assumptions on compiler: https://github.com/conda-forge/staged-recipes/pull/27490

File: recipes/ckzg/meta.yaml
Comment (MementoRC) [inline recipes/ckzg/meta.yaml:51]: This might not be needed

File: recipes/ckzg/meta.yaml
Comment (step21) [reply recipes/ckzg/meta.yaml]: +1

File: recipes/ckzg/meta.yaml
Comment (step21) [reply recipes/ckzg/meta.yaml]: true

**Comment by step21** [comment]
thanks for your comments. TBH I don't really care which PR we continue. Also added you as maintainer here. Though your has a lot more changes, I only patched out the build.bat so far, so I assume I would also have to carry those over.

**Comment by step21** [comment]
builds! all pkgs are using clang. Only thing is that there is a deprecation warning for setup.py, but I would ignore it for now.

**Comment by wolfv** [comment]
Looks good to me!

File: recipes/ckzg/meta.yaml
Comment (step21) [reply recipes/ckzg/meta.yaml:25]: I wasn't quite sure, but this matches what is recommended under cross-compilation in the knowledge base. See https://conda-forge.org/docs/maintainer/knowledge_base/#cross-compilation (you have to scroll a bit)

Comment (jakirkham) [reply recipes/ckzg/meta.yaml:25]: Usually we only need things that we execute (like `python`, compilers, etc.)

So we can probably drop `libblst` and `libblst-headers`

File: recipes/ckzg/conda_build_config.yaml
Comment (jakirkham) [inline recipes/ckzg/conda_build_config.yaml:8]: As we don't appear to use C++ here, we can drop this part

```suggestion
```

File: recipes/ckzg/meta.yaml
Comment (jakirkham) [inline recipes/ckzg/meta.yaml:26]: Suggesting dropping these per the discussion in this thread: https://github.com/conda-forge/staged-recipes/pull/26068#discussion_r1810520733

```suggestion
```

**Comment by step21** [comment]
Thanks @jakirkham for the suggestions, should all be applied now.

File: recipes/ckzg/meta.yaml
Comment (jakirkham) [inline recipes/ckzg/meta.yaml:32]: Appears `libblst` is missing `run_exports`, have raised this discussion in issue: https://github.com/conda-forge/libblst-feedstock/issues/2

Also it appears we don't have `libblst` in the global pinnings. So raised issue: https://github.com/conda-forge/conda-forge-pinning-feedstock/issues/6623

Would be good to sort these out so we can ensure this is being pinned appropriately

Comment (step21) [reply recipes/ckzg/meta.yaml:32]: ﻿Should be resolved once https://github.com/conda-forge/conda-forge-pinning-feedstock/pull/6655 is merged.

**Review by jakirkham — APPROVED** [review]

**Comment by step21** [comment]
Are we waiting for one more review now?

**Comment by jakirkham** [comment]
Right there were a bunch of folks looking at this. So was waiting to see if anyone else still had more thoughts

**Comment by step21** [comment]
Are we still waiting?

**Review by xhochy — APPROVED** [review]

**Comment by step21** [comment]
thanks!

**Pull request merged**
